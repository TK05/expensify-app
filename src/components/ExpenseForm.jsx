import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'


export default class ExpenseForm extends React.Component {
  static propTypes = {
    expense: PropTypes.shape({
      description: PropTypes.string,
      note: PropTypes.string,
      amount: PropTypes.string,
      createdAt: PropTypes.number
    }),
    onSubmit: PropTypes.func.isRequired,
    typeString: PropTypes.string.isRequired
  }

  static defaultProps = {
    expense: {},
  }

  constructor(props) {
    super(props)

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? props.expense.amount : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
      typeString: props.typeString
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({ note }))
  }

  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^[0-9]{1,}(\.[0-9]{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }

  onSubmit = (e) => {
    const {
      description, amount, note, createdAt
    } = this.state
    const { onSubmit } = this.props
    e.preventDefault()
    if (!description || !amount) {
      this.setState(() => ({
        error: 'Please provide description and amount'
      }))
    } else {
      this.setState(() => ({ error: '' }))
      onSubmit({
        description,
        amount: Number(amount).toFixed(2),
        note,
        createdAt: createdAt.valueOf()
      })
    }
  }

  render() {
    const {
      description, note, amount, createdAt, calendarFocused, error, typeString
    } = this.state
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {error && <p className="form__error">{error}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Description"
          value={description || ''}
          onChange={this.onDescriptionChange}
        />
        <input
          className="text-input"
          type="text"
          placeholder="Amount"
          value={amount || ''}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={createdAt}
          onDateChange={this.onDateChange}
          focused={calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          block
        />
        <textarea
          className="textarea"
          placeholder="Add a note for your expense (optional)"
          value={note}
          onChange={this.onNoteChange}
        />
        <div>
          <button className="button" type="submit">
            {`${typeString} Expense`}
          </button>
        </div>
      </form>
    )
  }
}
