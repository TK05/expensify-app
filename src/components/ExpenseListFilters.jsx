import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import 'react-dates/initialize'
import { DateRangePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'

import {
  setTextFilter as textFilter,
  sortByAmount as byAmount,
  sortByDate as byDate,
  setStartDate as startDateFunc,
  setEndDate as endDateFunc
} from '../actions/filters'


export class ExpenseListFilters extends React.Component {
  static propTypes = {
    text: PropTypes.string,
    sortBy: PropTypes.string.isRequired,
    startDateObj: PropTypes.objectOf(PropTypes.any),
    endDateObj: PropTypes.objectOf(PropTypes.any),
    setTextFilter: PropTypes.func.isRequired,
    sortByAmount: PropTypes.func.isRequired,
    sortByDate: PropTypes.func.isRequired,
    setStartDate: PropTypes.func.isRequired,
    setEndDate: PropTypes.func.isRequired
  }

  static defaultProps = {
    text: '',
    startDateObj: {},
    endDateObj: {}
  }

  state = {
    calendarFocused: null
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  }

  onTextChange = (e) => {
    const { setTextFilter } = this.props
    setTextFilter(e.target.value)
  }

  onSortChange = (e) => {
    const { sortByDate, sortByAmount } = this.props
    if (e.target.value === 'date') {
      sortByDate()
    } else if (e.target.value === 'amount') {
      sortByAmount()
    }
  }

  onDatesChange = ({ startDate, endDate }) => {
    const { setStartDate, setEndDate } = this.props
    setStartDate(startDate)
    setEndDate(endDate)
  }

  render() {
    const {
      text,
      sortBy,
      startDateObj,
      endDateObj
    } = this.props
    const { calendarFocused } = this.state

    return (
      <div>
        <input
          type="text"
          value={text}
          onChange={this.onTextChange}
        />
        <select
          value={sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={startDateObj}
          startDateId="startDate"
          endDate={endDateObj}
          endDateId="endDate"
          onDatesChange={this.onDatesChange}
          focusedInput={calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates
        />
      </div>
    )
  }
}

const mapStateToProps = state => ({
  text: state.filters.text,
  sortBy: state.filters.sortBy,
  startDateObj: state.filters.startDate,
  endDateObj: state.filters.endDate
})

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(textFilter(text)),
  sortByAmount: () => dispatch(byAmount()),
  sortByDate: () => dispatch(byDate()),
  setStartDate: dateObj => dispatch(startDateFunc(dateObj)),
  setEndDate: dateObj => dispatch(endDateFunc(dateObj))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
