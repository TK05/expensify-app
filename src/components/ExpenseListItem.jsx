import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import moment from 'moment'


const propTypes = {
  id: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  amount: PropTypes.string.isRequired,
  createdAt: PropTypes.number.isRequired,
}

export default function ExpenseListItem({
  id, description, amount, createdAt
}) {
  return (
    <Link className="list-item" to={`/edit/${id}`}>
      <div>
        <h3 className="list-item__title">{description}</h3>
        <span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
      </div>
      <h3 className="list-item__data">
        {new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(amount)}
      </h3>
    </Link>
  )
}

ExpenseListItem.propTypes = propTypes
