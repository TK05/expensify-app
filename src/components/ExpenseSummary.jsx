import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'

import selectExpenses from '../selectors/expenses'
import sumAllExpenses from '../selectors/expenses-total'

const propTypes = {
  expensesCount: PropTypes.number,
  expensesTotal: PropTypes.number
}

const defaultProps = {
  expensesCount: 0,
  expensesTotal: 0
}

export function ExpenseSummary({ expensesCount, expensesTotal }) {
  const expensesWord = expensesCount === 1 ? 'expense' : 'expenses'
  const formattedSum = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(expensesTotal)
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expensesCount}</span> {expensesWord} totalling <span>{formattedSum}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  )
}

ExpenseSummary.propTypes = propTypes
ExpenseSummary.defaultProps = defaultProps

const mapStateToProps = state => ({
  expensesCount: selectExpenses(state.expenses, state.filters).length,
  expensesTotal: sumAllExpenses(state.expenses)
})

export default connect(mapStateToProps)(ExpenseSummary)
