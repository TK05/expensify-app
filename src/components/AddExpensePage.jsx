import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../actions/expenses'


export class AddExpensePage extends React.Component {
  static propTypes = {
    addNewExpense: PropTypes.func.isRequired,
    history: PropTypes.objectOf(PropTypes.any).isRequired
  }

  onSubmit = (expense) => {
    const { addNewExpense, history } = this.props
    addNewExpense(expense)
    history.push('/')
  }

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          typeString="Add"
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  addNewExpense: expense => dispatch(startAddExpense(expense))
})

export default connect(null, mapDispatchToProps)(AddExpensePage)
