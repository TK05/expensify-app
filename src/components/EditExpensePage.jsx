import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'


export class EditExpensePage extends React.Component {
  static propTypes = {
    expense: PropTypes.objectOf(PropTypes.any).isRequired,
    editExpense: PropTypes.func.isRequired,
    removeExpense: PropTypes.func.isRequired,
    history: PropTypes.objectOf(PropTypes.any).isRequired
  }

  onSubmit = (editedExpense) => {
    const { editExpense, expense, history } = this.props
    editExpense(expense.id, editedExpense)
    history.push('/')
  }

  onRemove = () => {
    const { removeExpense, expense, history } = this.props
    removeExpense({ id: expense.id })
    history.push('/')
  }

  render() {
    const { expense } = this.props
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            typeString="Edit"
            expense={expense}
            onSubmit={this.onSubmit}
          />
          <button
            className="button button--secondary"
            type="button"
            onClick={this.onRemove}
          >
            Remove Expense
          </button>
        </div>
      </div>
    )
  }
}

const mapStateToProps = ({ expenses }, { match }) => ({
  expense: expenses.find(expense => expense.id === match.params.id)
})

const mapDispatchToProps = dispatch => ({
  editExpense: (id, updates) => dispatch(startEditExpense(id, updates)),
  removeExpense: id => dispatch(startRemoveExpense(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
