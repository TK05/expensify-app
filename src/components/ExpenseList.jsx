import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import selectExpenses from '../selectors/expenses'
import ExpenseListItem from './ExpenseListItem'


const propTypes = {
  expenses: PropTypes.arrayOf(PropTypes.object),
}

const defaultProps = {
  expenses: []
}

export function ExpenseList({ expenses }) {
  return (
    <div className="content-container">
      <div className="list-header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-desktop">Expense</div>
        <div className="show-for-desktop">Amount</div>
      </div>
      <div className="list-body">
        {
          expenses.length === 0 ? (
            <div className="list-item list-item--message">
              <span>No Expenses</span>
            </div>
          ) : (
            expenses.map(expense => (
              <ExpenseListItem
                key={expense.id}
                {...expense}
              />
            ))
          )
        }
      </div>
    </div>
  )
}

ExpenseList.propTypes = propTypes
ExpenseList.defaultProps = defaultProps

const mapStateToProps = state => (
  {
    expenses: selectExpenses(state.expenses, state.filters)
  }
)

export default connect(mapStateToProps)(ExpenseList)
