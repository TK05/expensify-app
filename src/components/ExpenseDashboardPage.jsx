import React from 'react'

import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseSummary from './ExpenseSummary'


export default function ExpenseDashboardPage() {
  return (
    <div>
      <ExpenseSummary />
      <ExpenseListFilters />
      <ExpenseList />
    </div>
  )
}
