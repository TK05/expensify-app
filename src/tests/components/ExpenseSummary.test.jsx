import React from 'react'
import { shallow } from 'enzyme'

import { ExpenseSummary } from '../../components/ExpenseSummary'

/* eslint-disable no-undef */
test('should render ExpenseSummary correctly with NO expenses', () => {
  const wrapper = shallow(
    <ExpenseSummary
      expensesLength={0}
      expensesTotal={0}
    />
  )
  expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseSummary correctly with ONE expense', () => {
  const wrapper = shallow(
    <ExpenseSummary
      expensesLength={1}
      expensesTotal={1234}
    />
  )
  expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseSummary correctly with expenses', () => {
  const wrapper = shallow(
    <ExpenseSummary
      expensesLength={10}
      expensesTotal={123456789.12}
    />
  )
  expect(wrapper).toMatchSnapshot()
})
