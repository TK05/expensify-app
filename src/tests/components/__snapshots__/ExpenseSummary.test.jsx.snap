// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly with NO expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totalling 
      <span>
        $0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary correctly with ONE expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totalling 
      <span>
        $1,234.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary correctly with expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totalling 
      <span>
        $123,456,789.12
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
